src/v/storage/segment_appender.cc

```diff
@@ -179,20 +173,17 @@ ss::future<> segment_appender::do_append(const char* buf, const size_t n) {
         }
     }
     if (written == n) {
-        return ss::make_ready_future<>();
+        co_return;
     }
 
-    return ss::get_units(_concurrent_flushes, 1)
-      .then([this, next_buf = buf + written, next_sz = n - written](
-              ssx::semaphore_units) {
-          // do not hold the units!
-          return internal::chunks().get().then(
-            [this, next_buf, next_sz](ss::lw_shared_ptr<chunk> chunk) {
-                vassert(!_head, "cannot overwrite existing chunk");
-                _head = std::move(chunk);
-                return do_append(next_buf, next_sz);
-            });
-      });
+    // barrier. do not hold the units!
+    auto units = co_await ss::get_units(_concurrent_flushes, 1);
+    units.return_all();
```

Comment:
I never did understand what this was there for in the original code. Do you?

Reply:
no--and it's even more confusing that it doesn't happen on the incoming path to the function. i'd say it's probably unnecessary, but haven't had the courage/motivation to remove it.